### PR TITLE
LDAP login support

### LINKED ISSUE #1396: Authenticate using LDAP
working on this with @ninanorgren 

https://github.com/Clinical-Genomics/scout/tree/ldap

### PR DESCRIPTION
**Pull request by northwestwitch**
This PR adds LDAP support to Scout (fix #1396) and fixes an annoying message error of the favicon.icon not found.

**How to test**:
1. App should be tested using 3 types of credentials:
- Normal user credentials stored in Scout mongodb database (can be done from local installation)
- Google credentials: should be done from Scout stage
- LDAP credentials. A bit tricky, detailed description below 

**First login type: credentials from database**:
- setup a demo instance of Scout and log in using the Clark Kent user (email: clark.kent@mail.com)
- You should be able to login normally

**Second login type: google credentials**:
- Install the branch on Scout stage and use your Google account to log in
- You should be able to login normally

**LDAP login**:
- Create a new user using the command line (had to modify it to include a new optional id input):
```
scout --demo load user -id newton -i cust000 -u "Isaac Newton" -m newton@fakeyemail.com
```
**Note that user _id in database does not have an email field.**

To test the LDAP access I've been using  the following online test server: https://www.forumsys.com/tutorials/integration-how-to/ldap/online-ldap-test-server/

To connect to this server you should edit the config file and add the following connection settings:
```
# LDAP connection settings
LDAP_PORT = 389
LDAP_HOST = 'ldap://ldap.forumsys.com'
LDAP_BASE_DN = 'dc=example,dc=com'
LDAP_BIND_USER_DN = 'cn=read-only-admin,dc=example,dc=com'
LDAP_REQUIRED_GROUP = 'ou=scientists,dc=example,dc=com'
LDAP_USER_SEARCH_SCOPE = 'SUBTREE',
LDAP_GROUP_OBJECT_FILTER = '(objectClass=GroupOfUniqueNames)'
```
Once server is restarted you should see a login page containing username and password fields.

You can test different login and password fields but the only ones working should be the combination:
```
username: newton
password: password
```

Take a screenshot and attach or copy/paste the output.

**Review:**
- [x] code approved by dNil
- [x] tests executed by dNil, CR

@ninanorgren could give it a try with your own ldap params when you have time?

### CODE REVIEW

**Comment by dnil** [comment]
Operational! Great job! I have not tested much more than what was in the spec, but I guess a bit of testing on the stuff that require and actual email, like validations or reruns, might be prudent too.

File: scout/server/blueprints/public/views.py [hunk 1]
```diff
@@ -13,8 +13,12 @@
...
 def index():
     """Show the static landing page."""
-
     form = None
     if current_app.config.get('LDAP_HOST'):
         form = LDAPLoginForm()
     return render_template('public/index.html', version=__version__, form=form)
+
+
+@public_bp.route('/favicon.ico')
```
Comment (dnil) [inline scout/server/blueprints/public/views.py]: 🥇

**Comment by dnil** [comment]
@ninanorgren: do let us know how it works in your hands! We are eager to hear, and looking forward to contributions in the future! 😸

**Comment by dnil** [comment]
Oh, and @northwestwitch: if you’d just update the Changelog please? Thanks!

**Comment by northwestwitch** [comment]
> Oh, and @northwestwitch: if you’d just update the Changelog please? Thanks!

Done! Thanks @dnil for reviewing and testing! 
Let's not merge immediately. It's better to confirm that it works with a real ldap server.

**Comment by dnil** [comment]
Ok, your call. I'd say it has been tested with a "real" remote server, so if it had been me I would have had it merged and fix any issues with a different server when the need arose..

**Comment by northwestwitch** [comment]
Ok, i'll just fix the code in a way that also the google login is testable and then we csn merge

**Comment by moonso** [comment]
Coveralls sais NO ⛔️

**Comment by northwestwitch** [comment]
I'm having a hard time testing the google login, but I fixed it in a way that many lines are moved out from the authorized function and tested elsewhere. The coverage went up so what about I test everything again after the latest commit and if it's working I'll merge?

**Comment by northwestwitch** [comment]
It's working! I've tested all 3 logins and the google one on stage. Merging this so we go forward with the other pull requests!

**Comment by moonso** [comment]
Great job @northwestwitch ! Ping @ninanorgren

**Comment by ninanorgren** [comment]
Sorry for the late reply, but now I've had time to test it a bit. Seems to work very well with the authentication. I just did a small test now but it seems it doesn't take REQUIRED_GROUP into consideration, it still logs people in although they don't belong to the group. Next week I have more time dedicated to this and I'll test it with more persons and see if it's the same problem. @northwestwitch @moonso

**Comment by northwestwitch** [comment]
> I just did a small test now but it seems it doesn't take REQUIRED_GROUP into consideration, it still logs people in although they don't belong to the group. 

Thanks for testing! Nice that it works, let us know if the group thing is a matter of config parameters of it must be fixed in the code!

**Comment by northwestwitch** [comment]
> 😊 Can't remember seeing any code for that? As far as I recall the discussion went into that you still need to add/white-list individual users, so no need to use groups on the LDAP? I suppose it would be possible to create new users on the fly if they match the group, and then remove them again if they try to log in later after having lost access to that group. That would be nifty, but as the base functionality for LDAP is now on par with what we have for google and local user based login, I think we'll leave that to you! 😉

Well yes, if the used id is not in the database the ldap user won't be able to connect.

**Comment by ninanorgren** [comment]
We discussed it a bit, and it was in the test app, but you're right, it's not super needed right now as we still have to enter people in the database. I can try to look at it a bit later, but not super high on my priority :) But super that the authentication works now, thank you @northwestwitch !

**Comment by northwestwitch** [comment]
> But super that the authentication works now

Nice to have you using Scout! :)

**Pull request merged**